pypiserver/_app.py

```diff
@@ -389,12 +391,29 @@ def json_info(project):
     if not packages:
         raise HTTPError(404, f"package {project} not found")
 
+    package_links = defaultdict(list)
+    for pkg in packages:
+        package_links[pkg.version].append(pkg.relfn_unix)
+    # links = [pkg.relfn_unix for pkg in packages]
+
     latest_version = packages[0].version
     releases = {}
     req_url = request.url
-    for x in packages:
-        releases[x.version] = [
-            {"url": urljoin(req_url, "../../packages/" + x.relfn)}
+    for package in packages:
+        matching_links = []
+        for version, links in package_links.items():
+            if version == package.version:
+                matching_links += links
+        releases[package.version] = [
+            {
+                "url": urljoin(req_url, "../../packages/" + link),
+                "digests": {
+                    "sha256": config.backend.digest_sha256(
+                        package, os.path.join(package.root, link)
+                    )
+                },
+            }
+            for link in matching_links
```

Comment:
Hey, just a wild thought got onto me here, do you think it would be possible to maintain `packages` and `package_links` as a single data structure? 🤔 

thinking of something along those lines:

```python
    packages = sorted(
        ...
    )

    # ...

    packages_with_links = [dict(package=pkg, version=pkg.version, link=pkg.relfn_unix) for pkg in packages]

    # ... and then working with `packages_with_links` in place of `matching_links` and `packages` nestedly?
```

But it's just a rough fantasy at this point, it sure needs working out better. What do you think?